### ISSUE TITLE
Custom Elements for Nav and Weekday

### ISSUE THREAD

**Issue by cwmoo740** [gpbl/react-day-picker#278]
The current syntax:

```jsx
<DayPicker weekdayElement={ <Weekday /> } navbarElement={ <Navbar /> } />
```

This doesn't allow using this library in Typescript in a typesafe way. In the code these elements are copied with React.cloneElement and additional props are added. Is there a reason for this instead of passing in the components themselves, e.g.

```jsx
function Navbar({previousMonth, nextMonth}) {}
class Weekday extends React.Component {}
<DayPicker weekdayElement={Weekday} navBarElement={Navbar} />
```

**Comment by gpbl** [comment]
In previous versions it was actually possible to pass classes to those attributes, but this feature has been deprecated in [2.3.0](https://github.com/gpbl/react-day-picker/blob/master/CHANGELOG.md#v230-2016-06-30).

The idea was to make easier to pass props to those components (see [this](https://github.com/gpbl/react-day-picker/pull/177)). Didn't know it would break the typescript way. 

What could be a viable solution?

**Comment by cwmoo740** [comment]
I think a common solution is to get people to write higher order components. This is the pattern that's encouraged by the react-redux connect function, for example.

This case is admittedly extra-complex because the extra properties you want to inject could be dynamic, but there are other ways to solve that - e.g. lodash memoize or this:

```
function AddProps(extraProps, OriginalComponent) {
    function WithExtraProps(props) {
        return (
            <OriginalComponent {props, ...extraProps} />
        );
    }
    WithExtraProps.displayName = `WithExtraProps(${OriginalComponent.name})`;
    return WithExtraProps;
}
function Weekday(props) {
    const {myProp, weekday} = props;
    return <div>weekday</div>;
}
class Example extends React.Component {
    constructor(props) {
        super(props);
        this.state = {hoc: AddProps({myProp: this.props.myWeekDayProp}, Weekday};
    }
    componentWillReceiveProps(nextProps) {
        if (this.props.myWeekDayProps !== nextProps.myWeekDayProps) {
            this.setState({hoc: AddProps({myProp: nextProps.myWeekDayProp}, Weekday);
        }
    }
    render() {
        return (
             <DayPicker weekday={this.state.hoc}/>;
        );
    }
}
```

Another idea is to support both an element and a component - 
```
if (WeekDayElement instanceof React.Component || typeof WeekDayElement === 'function') {
    return <WeekDayElement {...originalProps}/>
}
return React.cloneElement(WeekDayElement, originalProps);
```

**Comment by gpbl** [comment]
> Another idea is to support both an element and a component

yeah i was thinking about this solution. Could you send a PR to support `React.Component` classes as well? This change should apply to the other `*Element` props as well...

**Comment by cwmoo740** [comment]
Will do.

**Comment by cwmoo740** [comment]
https://github.com/gpbl/react-day-picker/pull/280

**Comment by gpbl** [comment]
Thanks, published as v5.2.0 🎉
